networking/eth2/src/main/java/tech/pegasys/teku/networking/eth2/rpc/beaconchain/methods/BlobSidecarsByRangeMessageHandler.java

```diff
@@ -203,50 +227,76 @@ class RequestState {
+    private Optional<Iterator<SlotAndBlockRootAndBlobIndex>> blobSidecarKeysIterator =
+        Optional.empty();
 
     RequestState(
         final ResponseCallback<BlobSidecar> callback,
         final UInt64 maxRequestBlobSidecars,
         final UInt64 startSlot,
-        final UInt64 endSlot) {
+        final UInt64 endSlot,
+        final Map<UInt64, Bytes32> canonicalHotRoots,
+        final UInt64 finalizedSlot) {
       this.callback = callback;
       this.maxRequestBlobSidecars = maxRequestBlobSidecars;
       this.startSlot = startSlot;
       this.endSlot = endSlot;
+      this.finalizedSlot = finalizedSlot;
+      this.canonicalHotRoots = canonicalHotRoots;
     }
 
     SafeFuture<Void> sendBlobSidecar(final BlobSidecar blobSidecar) {
       return callback.respond(blobSidecar).thenRun(sentBlobSidecars::incrementAndGet);
     }
 
     SafeFuture<Optional<BlobSidecar>> loadNextBlobSidecar() {
-      if (iterator.isEmpty()) {
+      if (blobSidecarKeysIterator.isEmpty()) {
         return combinedChainDataClient
             .getBlobSidecarKeys(startSlot, endSlot, maxRequestBlobSidecars)
             .thenCompose(
                 list -> {
-                  iterator = Optional.of(list.iterator());
-                  return getNextElement(iterator.get());
+                  blobSidecarKeysIterator = Optional.of(list.iterator());
+                  return getNextBlobSidecar(blobSidecarKeysIterator.get());
                 });
       } else {
-        return getNextElement(iterator.get());
+        return getNextBlobSidecar(blobSidecarKeysIterator.get());
       }
     }
 
-    private SafeFuture<Optional<BlobSidecar>> getNextElement(
-        final Iterator<SlotAndBlockRootAndBlobIndex> iterator) {
-      if (iterator.hasNext()) {
-        return combinedChainDataClient.getBlobSidecarByKey(iterator.next());
-      } else {
-        return SafeFuture.completedFuture(Optional.empty());
+    private SafeFuture<Optional<BlobSidecar>> getNextBlobSidecar(
+        final Iterator<SlotAndBlockRootAndBlobIndex> blobSidecarKeysIterator) {
+      if (blobSidecarKeysIterator.hasNext()) {
+        final SlotAndBlockRootAndBlobIndex slotAndBlockRootAndBlobIndex =
+            blobSidecarKeysIterator.next();
+
+        if (finalizedSlot.isGreaterThanOrEqualTo(slotAndBlockRootAndBlobIndex.getSlot())) {
+          return combinedChainDataClient.getBlobSidecarByKey(slotAndBlockRootAndBlobIndex);
+        }
+
+        // not finalized, let's check if it is on canonical chain
+        if (isCanonical(slotAndBlockRootAndBlobIndex)) {
+          return combinedChainDataClient.getBlobSidecarByKey(slotAndBlockRootAndBlobIndex);
+        }
+
+        // non-canonical, try next one
+        return getNextBlobSidecar(blobSidecarKeysIterator);
       }
+
+      return SafeFuture.completedFuture(Optional.empty());
+    }
+
+    private boolean isCanonical(final SlotAndBlockRootAndBlobIndex slotAndBlockRootAndBlobIndex) {
```

Comment:
I'd clarify, something like `isCanonicalHotBlobSidecar`, maybe you get something shorter on a mind